### ISSUE TITLE
Detail Problem

### ISSUE THREAD

**Issue by yyEudora** [omarfoq/knn-per#6]
Dear researcher, 
      I have read your paper recently and appreciate your research results very much. I tried to reproduce the results presented in your paper, but I still cannot get the result of Figure 1, and I don't know where the problem is. Could you  please write the document in more detail? 
      In addition, I would like to ask why columns 2, 10, 25 and -1 are selected by cifar10 in Figure 2 as the test accuracy results of the local storage sizes of 5, 25, 100 and 250 respectively. Isn't average_scores output the average accuracy of 200 clients under the combination of capacity and weight? If you select a column, it is the accuracy of the fixed weight parameter, and how does it relate to the size of the local data set?

**Comment by yyEudora** [comment]
One more question, is the alpha in the pre-trained model the same as the alpha when the data is generated? Looking forward to your answer. Thank you!

**Comment by omarfoq** [comment]
Hello @yyEudora 

Thank you for your message. Regarding the reproduction of Figure 1, would you please let me know what parts of the figure you cannot reproduce, or what are the results you obtained? Also, could you please let me know if you are using the pre-trained model available [here](https://drive.google.com/drive/folders/1lu0XdKO7GRtQdU99p6matevddU9TOr__) or if you are retraining the model from scratch? 

Regarding `average_scores` and Figure 2, the shape of `average_scores` should be `(len(weights_grid_), len(capacities_grid_))` (as you can see in [eval_knnper.py#L6](https://github.com/omarfoq/knn-per/blob/main/eval_knnper.py#L6)). Therfore, selecting a column is equivalent to sleecting the average accuracy across client at a fixed capacity value. The capacity is equal to the number of allowed samples for prediction divided by the total number of local samples. 

The alpha in pre-trained model is the same as the alpha when the data is generated. 

I hope I answer your (second and third) questions, please let me know if you have any further question.

**Comment by yyEudora** [comment]
Dr. Marfoq:
Hello! Thank you for your answer, I have a feeling of clarity after reading your answer today.
I have a new question about the use of KNN:
By further understanding the code of your paper, I found that you have extracted the features from the training data by the global model, saved the corresponding (features, labels) locally, and then performed the next knn operation on the test data. So does this approach consume more local memory overhead compared to storing the model parameters? Is it a space-for-accuracy approach? If I replace the feature extractor from the global model to the client model (personalized model), is this approach feasible or reasonable?

**Comment by omarfoq** [comment]
Hello, 

I think it is a space-for-computation, instead of space-for-accuracy, approach. Our implementation choice consumes more memory in comparison to only storing the model parameters  as you mentioned. But does not require any significant computation at inference time. 

Regarding the second part of your question, I think it is related to issue #5. Please let me know if the answer therein does not address your question.

**Comment by yyEudora** [comment]
Thank you very much for your enthusiastic reply, your answer really cleared up my doubts. Regarding your answer in issue #5, I have also verified it experimentally and the results are consistent with what you have guessed. Thank you very much!

> Hello,
> 
> I think it is a space-for-computation, instead of space-for-accuracy, approach. Our implementation choice consumes more memory in comparison to only storing the model parameters as you mentioned. But does not require any significant computation at inference time.
> 
> Regarding the second part of your question, I think it is related to issue #5. Please let me know if the answer therein does not address your question.
